### ISSUE TITLE
Possible to use SQLITE_FCNTL_PERSIST_WAL?

### ISSUE THREAD

**Issue by nielspeen** [rqlite/rqlite#1902]
Is it possible/safe to use [SQLITE_FCNTL_PERSIST_WAL](https://www.sqlite.org/c3ref/c_fcntl_begin_atomic_write.html#sqlitefcntlpersistwal) with rqlite?

I'm aware that directly reading the sqlite file is not officially supported, but it works quite well when taking the necessary precautions. The only problem that occasionally pops-up is when a read takes place during the brief moment that the WAL file is deleted and re-created. It's my understanding that SQLITE_FCNTL_PERSIST_WAL resolves this by ensuring a WAL file is always present.

**Comment by otoolep** [comment]
The short answer is no, do not nothing that changes the configuration of the SQLite database underneath rqlite. See https://rqlite.io/docs/guides/direct-access/

As for that flag that is not what `SQLITE_FCNTL_PERSIST_WAL` does actually AFAIK. To quote https://www.sqlite.org/draft/c3ref/c_fcntl_begin_atomic_write.html

_Setting persistent WAL mode causes those files to persist after close._

In other words it stops SQLite from deleting the WAL on close, even after checkpointing it.

rqlite never deletes the WAL file during operation, but it does truncate it to zero length (CHECKPOINT TRUNCATE) periodically during checkpointing (though maybe there is some corner case in the code I've forgotten about). It does delete the SQLite files on startup. (This will change in rqlite 9.0 however).

Is there some particular issue you're hitting? Maybe I can help you address it. But because rqlite is using standard SQLite commands to modify the WAL, I'm surprised it's causing your readers any problems.

**Comment by otoolep** [comment]
In other words, there should always be a WAL file present, except at rqlite restart time.

**Comment by nielspeen** [comment]
When opening SQLite in READ_ONLY mode it will occasionally try to write to the WAL. Allowing this causes corruption, so I'm using filesystem permissions to prevent this.

I don't know why it wants to write to the WAL. I assume because it's either gone or in a non-usable state for a brief moment. 

My conclusion and suggestion are based on reading this thread: https://github.com/groue/GRDB.swift/issues/771 and this comment in particular: https://github.com/groue/GRDB.swift/issues/771#issuecomment-624441052

**Comment by otoolep** [comment]
**TL;DR: run the latest version of rqlite. I do not of any reason you shouldn't be able to read the database using a client that only has read-only permissions.**

I'm not entirely following your last point.

Are you saying you want the WAL and -shm to always persist, so that you can open a read-only connection to a WAL enabled SQLite file? Have you read section 5 of https://www.sqlite.org/draft/wal.html ? 

rqlite does two things to ensure read-only connections of the underlying database should work:

- it doesn't checkpoint the WAL file when it shuts down. Therefore the WAL file should remain on disk even if rqlite is not running.
- rqlite explicitly creates the -shm and -wal files immediately when it starts, regardless of whether any writes are sent to rqlite. This change took effect in v8.30.1.

So be sure you're running the latest release. As long as `rqlited` is running I don't know of any blockers to reading the database from another process. If `rqlited` is not running however, that is not something I can do anything about though (as far as I know). It might work, it might now.

**Comment by otoolep** [comment]
>The only problem that occasionally pops-up is when a read takes place during the brief moment that the WAL file is deleted and re-created. 

When exactly is this? rqlite, if running the latest release, never deletes the WAL file, it only truncates it to zero length. But the file  itself never goes away. That's what I don't understand about your request, unless you are running old code.

**Comment by nielspeen** [comment]
Running 8.30.1 currently. Will update to 8.30.4 today.

For reference:

$ ps -eo pid,comm,stime | grep rqlited
 329112 rqlited         **Sep10**

$ stat db.sqlite-wal 
  File: db.sqlite-wal
  Size: 226632    	Blocks: 448        IO Block: 4096   regular file
Device: 1ch/28d	Inode: 154654118   Links: 1
Access: (0644/-rw-r--r--)  Uid: ( 1001/  rqlite)   Gid: ( 1001/  rqlite)
Access: 2024-09-22 04:24:38.448119482 +0800
Modify: 2024-09-22 04:24:40.756223519 +0800
Change: 2024-09-22 04:24:40.756223519 +0800
Birth: **2024-09-21** 07:26:22.068325947 +0800

**Comment by nielspeen** [comment]
> When exactly is this? rqlite, if running the latest release, never deletes the WAL file, it only truncates it to zero length. But the file itself never goes away. That's what I don't understand about your request, unless you are running old code.

Thanks. This confused me for a bit, as ```stat``` was giving me _birth_ times for the WAL file that were more recent than the start of rqlite. But it turns out those coincide exactly with the auto-vacuum.

I'll have to take a few steps back to see where the read issue comes from. For reference, the error I will get is _General error: 8 attempt to write a readonly database._ It's a pretty rare error. Perhaps once every 24-48 hours on a very busy cluster.

**Comment by otoolep** [comment]
OK, let me know.

**Comment by otoolep** [comment]
Automatic vacuum involves writing the SQLite to a brand new file, and then rqlite opens that one instead. Maybe your client is getting confused by that.

**Comment by otoolep** [comment]
I'm going to close this for now, please re-open with your errors if you hit them.

**Comment by otoolep** [comment]
>Automatic vacuum involves writing the SQLite to a brand new file, and then rqlite opens that one instead. Maybe your client is getting confused by that.

@nielspeen -- yeah, from looking at the code, perhaps that is it. Since AutoVac does recreate the database file and there is a very small window where -shm and -wal files would also be an inconsistent state (or may not exist) this may be causing your client confusion.

I'm actually going to look into changing auto-vac so this file-recreation doesn't happen -- from looking at the code it may not actually be necessary.. Stay tuned.

Again, not sure it's the cause, but just in case.

**Comment by otoolep** [comment]
@nielspeen -- you should upgrade to https://github.com/rqlite/rqlite/releases/tag/v8.31.2

I can't be sure, but it's possible the issues you were seeing were related to how AutoVac worked. In the latest release the SQLite file is no longer recreated on the fly, possibly while your client was reading them.
